spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -172,171 +149,69 @@ public RedisLockRegistry(RedisConnectionFactory connectionFactory, String regist
-	 */
-	private Collection<RedisLock> getHardThreadLocks() {
-		List<RedisLock> locks = this.hardThreadLocks.get();
-		if (locks == null) {
-			locks = new LinkedList<RedisLock>();
-			this.hardThreadLocks.set(locks);
-		}
-		return locks;
-	}
-
-	private RedisLock findLock(Collection<RedisLock> locks, Object key) {
-		if (locks != null) {
-			for (RedisLock lock : locks) {
-				if (lock.getLockKey().equals(key)) {
-					return lock;
-				}
-			}
-		}
-		return null;
-	}
-
-	private void toHardThreadStorage(RedisLock lock) {
-		if (this.weakThreadLocks.get() != null) {
-			this.weakThreadLocks.get().remove(lock);
-		}
-
-		getHardThreadLocks().add(lock);
-
-		//clean up
-		if (this.weakThreadLocks.get() != null && this.weakThreadLocks.get().isEmpty()) {
-			this.weakThreadLocks.remove();
-		}
-	}
-
-	private void toWeakThreadStorage(RedisLock lock) {
-		//to avoid collection creation on existence check use direct fields
-		if (this.hardThreadLocks.get() != null) {
-			getHardThreadLocks().remove(lock);
-		}
-
-		if (this.useWeakReferences) {
-			getWeakThreadLocks().add(lock);
-		}
-
-		//clean up
-		if (this.hardThreadLocks.get() != null && this.hardThreadLocks.get().isEmpty()) {
-			this.hardThreadLocks.remove();
-		}
 	}
 
 	@Override
 	public Lock obtain(Object lockKey) {
 		Assert.isInstanceOf(String.class, lockKey);
-
-		//try to find the lock within hard references
-		RedisLock lock = findLock(this.hardThreadLocks.get(), lockKey);
-
-		/*
-		 * If the lock is locked, check that it matches what's in the store.
-		 * If it doesn't, the lock must have expired.
-		 */
-		if (lock != null && lock.thread != null) {
-			RedisLock lockInStore = this.redisTemplate.boundValueOps(this.registryKey + ":" + lockKey).get();
-			if (lockInStore == null || !lock.equals(lockInStore)) {
-				try {
-					lock.unlock();
-				}
-				catch (Exception e) {
-					if (logger.isWarnEnabled()) {
-						logger.warn("Lock was released due to expiration. A new one will be obtained...", e);
-					}
-				}
-				if (this.hardThreadLocks.get() != null) {
-					this.hardThreadLocks.get().remove(lock);
-				}
-				if (this.weakThreadLocks.get() != null) {
-					this.weakThreadLocks.get().remove(lock);
-				}
-				lock = null;
+		String path = (String) lockKey;
+		RedisLock lock;
+		synchronized (this.locks) {
+			lock = this.locks.get(path);
+			if (lock == null) {
+				lock = new RedisLock(path);
+				this.locks.put(path, lock);
 			}
 		}
+		return lock;
+	}
 
-		if (lock == null) {
-			//try to find the lock within weak references
-			lock = findLock(this.weakThreadLocks.get(), lockKey);
-
-			if (lock == null) {
-				lock = new RedisLock((String) lockKey);
-
-				if (this.useWeakReferences) {
-					getWeakThreadLocks().add(lock);
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		synchronized (this.locks) {
+			Iterator<Map.Entry<String, RedisLock>> iterator = this.locks.entrySet().iterator();
```

Comment:
`ExpirableLockRegistry`:
```
* A {@link LockRegistry} implementing this interface supports the removal of aged locks
* that are not currently locked.
```
Since we can't assume if the lock in the Redis is locked in some other process, we can remove only local instances.
And from the [Reference Manual](http://docs.spring.io/spring-integration/reference/html/zookeeper.html#zk-lock-registry):

> For unbounded strategies (such as the default) you will need to invoke expireUnusedOlderThan(long age) from time to time, to remove old unused locks from memory.

So, we are really talking here only about **in-memory**.